### DISCUSSION TITLE
How to Prompt the User to Choose a PDF File Storage Location in Blazor using QuestPDF

### DISCUSSION THREAD

**Discussion by aidrecabrera** [QuestPDF/QuestPDF#692] [Q&A] [answered]
I'm working on a Blazor application and using the QuestPDF library to generate PDFs. I want to provide a feature where, when the user clicks a button, they can choose the location where the generated PDF file will be stored. I'm looking for guidance on how to implement this in Blazor using QuestPDF.

When the user clicks a specific button, I want to trigger a file dialog that allows the user to select the directory or location where the PDF file should be saved.

1. How can I show a file dialog in a Blazor application to let the user choose the directory for PDF file storage?
2. Are there any specific QuestPDF features or functions that can assist in this process?

Any help, advice, or code examples would be greatly appreciated.

Thank you!

**Comment by girlpunk** [comment, marked as answer]
This is fairly out-of-scope for the QuestPDF library, however you could try something like this:

```csharp
var file = Document.GeneratePdf();

using var streamRef = new DotNetStreamReference(new MemoryStream(file));
await JSRuntime.InvokeVoidAsync(
        "downloadFileFromStream",
        $"{filename}.pdf",
        streamRef
    )
    .ConfigureAwait(false);
```

```javascript
async function downloadFileFromStream(fileName, contentStreamReference) {
    const arrayBuffer = await contentStreamReference.arrayBuffer();

    const blob = new Blob([arrayBuffer], { type: "application/pdf" });
    const url = URL.createObjectURL(blob);

    const anchorElement = document.createElement('a');
    anchorElement.href = url;

    if (fileName) {
        anchorElement.download = fileName;
    }

    anchorElement.target = "_blank";
    document.body.appendChild(anchorElement);
    anchorElement.innerText = "Click to download!";

    anchorElement.click();
    document.body.removeChild(anchorElement);
    anchorElement.remove();

    URL.revokeObjectURL(url);
}
```

**Reply by aidrecabrera** [reply]
Thank you for this! This is the solution that I was looking for.

**Reply by MarcinZiabek** [reply]
I fully agree that this feature is completely out-of-scope for the library. 

@girlpunk Thank you for providing such a great example!

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
